include/data_sink.hpp

```diff
@@ -0,0 +1,900 @@
+    }
+};
+
+namespace detail {
+template<typename T, typename P>
+std::span<T>
+find_matching_prefix(std::span<T> s, P predicate) {
+    const auto nm = std::find_if_not(s.begin(), s.end(), predicate);
+    if (nm == s.end()) {
+        return s;
+    }
+    return s.first(std::distance(s.begin(), nm));
+}
+
+template<typename T>
+bool
+copy_span(std::span<const T> src, std::span<T> dst) {
+    assert(src.size() <= dst.size());
+    if (src.size() > dst.size()) {
+        return false;
+    }
+    std::copy(src.begin(), src.end(), dst.begin());
+    return true;
+}
+} // namespace detail
+
+/**
+ * @brief generic data sink for exporting arbitrary-typed streams to non-GR C++ APIs.
+ *
+ * Each sink registers with a (user-defined/exchangeable) global registry that can be
+ * queried by the non-GR caller to find the sink responsible for a given signal name, etc.
+ * and either retrieve a poller handler that allows asynchronous data from a different thread,
+ * or register a callback that is invoked by the sink if the user-conditions are met.
+ *
+ * <pre>
+ * @code
+ *         ╔═══════════════╗
+ *    in0 ━╢   data sink   ║                      ┌──── caller ────┐
+ * (err0) ━╢ (opt. error)  ║                      │                │
+ *         ║               ║  retrieve poller or  │ (custom non-GR │
+ *        ━╢ :signal_name  ║←--------------------→│  user code...) │
+ *        ━╢ :signal_unit  ║  register            │                │
+ *         ║ :...          ║  callback function   └───┬────────────┘
+ *         ╚═ GR block ═╤══╝                          │
+ *                      │                             │
+ *                      │                             │
+ *                      │      ╭─registry─╮           │
+ *            register/ │      ╞══════════╡           │ queries for specific
+ *          deregister  ╰─────→│ [sinks]  │←──────────╯ signal_info_t list/criteria
+ *                             ╞══════════╡
+ *                             ╰──────────╯
+ *
+ * </pre>
+ * Pollers can be configured to be blocking, i.e. blocks the flow-graph
+ * if data is not being retrieved in time, or non-blocking, i.e. data being dropped when
+ * the user-defined buffer size is full.
+ * N.B. due to the nature of the GR scheduler, signals from the same sink are notified
+ * synchronuously (/asynchronuously) if handled by the same (/different) sink block.
+ *
+ * @tparam T input sample type
+ */
+template<typename T>
+class data_sink : public node<data_sink<T>> {
+    struct abstract_listener;
+
+    static constexpr std::size_t                   _listener_buffer_size = 65536;
+    std::deque<std::unique_ptr<abstract_listener>> _listeners;
+    std::mutex                                     _listener_mutex;
+    gr::history_buffer<T>                          _history = gr::history_buffer<T>(1);
+
+public:
+    Annotated<float, "sample rate", Doc<"signal sample rate">, Unit<"Hz">>           sample_rate = 10000.f;
```

Comment:
maybe a default of `1` might be better?!? This would correspond to a horizontal axis of 1 tick == 1 sample.
Similar for signal_unit which should maybe default to `a.u.` or `Ts`?

signal_min and signal_max should be at the min/max rails of the type `T`

signal_name default value should be either `unknown_siganal` or maybe through a static (atomic?) global counter `i` something like `unknown_signal#<i>`?

Reply:
Added the defaults, as well as reading the values from settings and tags (settings have preference), and forwarding the info to the datasets.